src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -1532,6 +1532,30 @@ theorem signExtend_eq_setWidth_of_lt (x : BitVec w) {v : Nat} (hv : v ≤ w):
 theorem signExtend_eq (x : BitVec w) : x.signExtend w = x := by
   rw [signExtend_eq_setWidth_of_lt _ (Nat.le_refl _), setWidth_eq]
 
+theorem toNat_signExtend_of_gt (x : BitVec w) {v : Nat} :
+  (x.signExtend (w + v)).toNat = x.toNat + (2 ^ (w + v)  - 2^w) * (if x.msb then 1 else 0) := by
+  apply Nat.eq_of_testBit_eq
+  intro i
+  rw [testBit_toNat, getLsbD_signExtend, Nat.pow_add, ← Nat.mul_sub_one,
+      Nat.add_comm (x.toNat), Nat.mul_assoc, Nat.testBit_mul_pow_two_add _ (x.isLt)]
+  by_cases h : i < w
+  · simp [h, Nat.lt_add_right v h, testBit_toNat]
+  · by_cases h2 : x.msb
+    <;> by_cases h1 : i < w + v
+    <;> simp [h2, h, h1, Nat.sub_lt_left_of_lt_add (Nat.le_of_not_lt h)]
+    exact Nat.le_sub_of_add_le (Nat.add_comm _ _ ▸ (Nat.le_of_not_lt h1))
+
+theorem toNat_signExtend (x : BitVec w) {v : Nat} :
+  (x.signExtend v).toNat = (x.setWidth v).toNat + (2 ^ v - 2 ^ w) * (if x.msb then 1 else 0) := by
+  by_cases h : v ≤ w
```

Comment:
```suggestion
  (x.signExtend v).toNat = (x.setWidth v).toNat + if x.msb then (2 ^ v - 2 ^ w) else 0#w := by
  by_cases h : v ≤ w
```

Would this be more readable?

Reply:
Also, why does setWidth reappear on the RHS?

Reply:
If v < w, then signExtend is just setWidth/truncation.

Reply:
What did we decide here?